File: apps/android-companion/app/src/test/java/com/nova/companion/CompanionCryptoTest.kt
package com.nova.companion

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.security.KeyPairGenerator
import java.security.spec.ECGenParameterSpec
import java.util.Base64

class CompanionCryptoTest {

    private fun generateKeyPairB64(): Pair<String, String> {
        val pair = KeyPairGenerator.getInstance("EC").apply { initialize(ECGenParameterSpec("secp256r1")) }.generateKeyPair()
        return Base64.getEncoder().encodeToString(pair.public.encoded) to Base64.getEncoder().encodeToString(pair.private.encoded)
    }

    @Test
    fun `both sides of an ECDH exchange derive the identical session key`() {
        val (desktopPub, desktopPriv) = generateKeyPairB64()
        val (mobilePub, mobilePriv) = generateKeyPairB64()

        val desktopSideKey = CompanionCrypto.deriveSessionKey(desktopPriv, mobilePub)
        val mobileSideKey = CompanionCrypto.deriveSessionKey(mobilePriv, desktopPub)

        assertArrayEquals(desktopSideKey, mobileSideKey)
        assertEquals(32, desktopSideKey.size)
    }

    @Test
    fun `round-trips a payload encrypted and decrypted with the same derived key`() {
        val (desktopPub, desktopPriv) = generateKeyPairB64()
        val (mobilePub, mobilePriv) = generateKeyPairB64()
        val key = CompanionCrypto.deriveSessionKey(desktopPriv, mobilePub)
        val peerKey = CompanionCrypto.deriveSessionKey(mobilePriv, desktopPub)

        val envelope = CompanionCrypto.encrypt(key, "{\"hello\":\"nova\"}")
        val decrypted = CompanionCrypto.decrypt(peerKey, envelope)

        assertEquals("{\"hello\":\"nova\"}", decrypted)
    }

    @Test
    fun `rejects a payload decrypted with the wrong key`() {
        val (pub1, priv1) = generateKeyPairB64()
        val (pub2, _) = generateKeyPairB64()
        val (pub3, priv3) = generateKeyPairB64()
        val key = CompanionCrypto.deriveSessionKey(priv1, pub2)
        val wrongKey = CompanionCrypto.deriveSessionKey(priv3, pub2)
        assertTrue(!key.contentEquals(wrongKey))
        assertTrue(pub1 != pub3) // sanity: the two generated key pairs are actually distinct

        val envelope = CompanionCrypto.encrypt(key, "secret")
        val result = CompanionCrypto.decrypt(wrongKey, envelope)

        assertNull(result)
    }
}

File: services/runtime/test/companion-crypto.test.ts
import { describe, expect, it } from "vitest";

import {
  decryptPayload,
  deriveSessionKey,
  encryptPayload,
  generateCompanionKeyPair,
  signChallenge,
  verifyChallengeSignature,
} from "../src/companion-crypto.js";

describe("companion-crypto", () => {
  it("both sides of an ECDH exchange derive the identical session key", () => {
    const desktop = generateCompanionKeyPair();
    const mobile = generateCompanionKeyPair();

    const desktopSideKey = deriveSessionKey(desktop.privateKeyB64, mobile.publicKeyB64);
    const mobileSideKey = deriveSessionKey(mobile.privateKeyB64, desktop.publicKeyB64);

    expect(desktopSideKey.ok).toBe(true);
    expect(mobileSideKey.ok).toBe(true);
    if (!desktopSideKey.ok || !mobileSideKey.ok) return;
    expect(desktopSideKey.value.equals(mobileSideKey.value)).toBe(true);
    expect(desktopSideKey.value.length).toBe(32);
  });

  it("round-trips a payload encrypted and decrypted with the same derived key", () => {
    const desktop = generateCompanionKeyPair();
    const mobile = generateCompanionKeyPair();
    const key = deriveSessionKey(desktop.privateKeyB64, mobile.publicKeyB64);
    expect(key.ok).toBe(true);
    if (!key.ok) return;

    const envelope = encryptPayload(key.value, JSON.stringify({ hello: "nova" }));
    const decrypted = decryptPayload(key.value, envelope);

    expect(decrypted).toMatchObject({ ok: true, value: JSON.stringify({ hello: "nova" }) });
  });

  it("rejects a payload decrypted with the wrong key", () => {
    const pair = generateCompanionKeyPair();
    const otherPair = generateCompanionKeyPair();
    const key = deriveSessionKey(pair.privateKeyB64, otherPair.publicKeyB64);
    const wrongKey = deriveSessionKey(
      otherPair.privateKeyB64,
      generateCompanionKeyPair().publicKeyB64,
    );
    expect(key.ok && wrongKey.ok).toBe(true);
    if (!key.ok || !wrongKey.ok) return;

    const envelope = encryptPayload(key.value, "secret");
    const result = decryptPayload(wrongKey.value, envelope);

    expect(result.ok).toBe(false);
  });

  it("verifies a signature produced with the matching private key and rejects a tampered challenge", () => {
    const pair = generateCompanionKeyPair();
    const challenge = Buffer.from("nova-pairing-challenge");
    const signature = signChallenge(pair.privateKeyB64, challenge);

    expect(verifyChallengeSignature(pair.publicKeyB64, challenge, signature)).toBe(true);
    expect(verifyChallengeSignature(pair.publicKeyB64, Buffer.from("tampered"), signature)).toBe(
      false,
    );
  });
});
